drivers: ipm: Added IPM nofitication over IVSHMEM Doorbell.

uLipe opened the pull request:
This PR aims to add an IPM device driver based on top of the IVSHMEM Doorbell notification present in the IVSHMEM driver, the idea is to have an common way to provide a cross-core (cross-instance in this case) interrupt between two QEMU running Zephyr, this mechanism can be used as primitive to support for example RPMsg service for ARM64 architectures that don't have shared memory plus a cross core notification method.

uLipe commented:
@wmamills you might be interested on this as well.

carlocaione reviewed (changes requested):
But why an IPM driver and not an MBOX driver instead? MBOX has better DT support, better functionality, it supports multiple channels and it supports IPC subsystem out of the box. IPM is almost legacy at this point.

Do you any strong requirement to support this using IPM instead of MBOX?

uLipe commented:
The main reason to use IPM instead of the MBOX driver is because rpmsg_service in Zephyr currents expects and IPM driver instead of the MBOX, as well other oenamp samples. 

RPMsg protocol (as well other openamp protocols) expects from a cross-core notification driver in the guest OS, IPM in this case.

Of course this driver can be converted into a MBOX one but doing this right now would make the usage of something like RPMsg over IVSHMEM among `qemu_cortex_a53` instances to not work out of the box.

carlocaione commented:
> The main reason to use IPM instead of the MBOX driver is because rpmsg_service in Zephyr currents expects and IPM driver instead of the MBOX, as well other oenamp samples.

Right, but also `rpmsg_service` is being super-seeded by IPC service with the static VRINGs backend.

uLipe commented:
I can provide the both drivers in that case, and deprecate the IPM driver when openamp sample and rpmsg_service gets deprecated or superseeded by the MBOX driver, although for now to use openamp over qemu istances I find the IPM driver is still required.

carlocaione commented:
> I can provide the both drivers in that case, and deprecate the IPM driver when openamp sample and rpmsg_service gets deprecated or superseeded by the MBOX driver, although for now to use openamp over qemu istances I find the IPM driver is still required.

Providing the same driver also for MBOX would make me an happy person :heart:

uLipe commented:
The failed jobs are related to already existing warnings on the IVSHMEM shell, should I fix that in this PR?

carlocaione commented on drivers/ipm/Kconfig:
```diff
@@ -38,10 +38,20 @@ config ESP32_SOFT_IPM
+config IPM_IVSHMEM
+	bool "IPM driver based on IVSHMEM-Doorbell"
+	default y
+	depends on IVSHMEM
+	depends on IVSHMEM_DOORBELL
```
where is the `depends on DT_HAS_..`?

carlocaione commented on drivers/ipm/ipm_ivshmem.c:
```diff
@@ -0,0 +1,133 @@
+
+struct ivshmem_ipm_data {
+	ipm_callback_t cb;
+    void *user_data;
```
oddly indented

carlocaione commented on drivers/ipm/ipm_ivshmem.c:
```diff
@@ -0,0 +1,133 @@
+
+    const struct device *dev = (const struct device *)arg;
+    struct ivshmem_ipm_data *dev_data = (struct ivshmem_ipm_data *)dev->data;
+    struct ivshmem_ipm_config *dev_cfg = (struct ivshmem_ipm_config *)dev->config;
```
again, oddly indented.

carlocaione reviewed (changes requested):
All the indentation changes in commit 3 should be squashed in the first commit.

Commit:
drivers: ipm: added IPM over IVSHMEM driver

This driver is built on top of the IVSHMEM doorbell
notification mechanism providing an unified way
to generate inter VM interrupts.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

Commit:
samples: drivers: ipm: added IPM over IVSHMEM sample

To demonstrate how to configure Zephyr to use the IPM
driver over the IVSHMEM subsystem. Since this driver
is intended to generate inter QEMU VM notifications
it was better to create a sample app using the shell
for quick demonstration for the users.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

carlocaione commented on drivers/ipm/ipm_ivshmem.c:
```diff
@@ -0,0 +1,135 @@
+	int ret = ivshmem_register_handler(dev_cfg->ivshmem_dev, &sig, 0);
+
+	if (ret < 0) {
+		LOG_ERR("registering handlers must be supported: %d\n", ret);
+		k_panic();
```
do you really need to panic here?

uLipe replied:
From what I know of the IVSHMEM driver, failure from registering handlers is a unrecoverable error.

carlocaione commented on drivers/ipm/ipm_ivshmem.c:
```diff
@@ -0,0 +1,135 @@
+	return 0;
+}
+
+K_THREAD_STACK_DEFINE(ivshmem_ev_loop_stack, CONFIG_IPM_IVSHMEM_EVENT_LOOP_STACK_SIZE);
+static struct k_thread ivshmem_ev_loop_thread;
```
move this after the headers.

carlocaione commented on drivers/ipm/Kconfig.ivshmem:
```diff
@@ -0,0 +1,18 @@
+	help
+	  Adjust the stack size, in bytes of the ivshmem event loop task.
+
+config IPM_IVSHMEM_EVENT_LOOP_PRIO
+	int "Priority of IVSHMEM IPM Event loop task"
+	default 2
```
Why do you need such high cooperative priority? Cooperative tasks must really be carefully justified because can easily starve the scheduler if used not appropriately.

uLipe replied:
It is just a way to emulate the cross-core interrupt service, as other IPM drivers (or MBOX) conventions, heavy work should not be done there.

carlocaione replied:
The problem is that in the cooperative thread you are going to call `dev_data->cb`, that means that the user is basically installing an handler in a very high priority cooperative thread without even knowing it.

Cooperative threads are dangerous because if the user doesn't know what he is doing or even that the code that he installed is in a cooperative threads, things can go south very quickly.

Think about someone installing in the callback a busy-loop of some kind, that means that the scheduler is basically hostage of the user that is DoS-sing the whole platform.

As someone said: if you feel the need to use a cooperative thread, you are probably wrong.

uLipe replied:
I understand the implication of usage the cooperative threads without caution, my point is more from how the user perceiveis the consumption of the IPM driver, and it is documented as follow:

carlocaione replied:
I see but I still do not see why a cooperative thread must be used here. Yes, you are more compliant to the API but at what cost? You are offloading too much responsibility to the user IMO for what gain? I can understand the usage of cooperative threads when you really, really want to be sure to not be interrupted, but if you do not have that kind of need or some kind of performance constrain the usage of cooperative threads is to be avoided as much as possible.

Please note that I'm suggesting this because I was personally heavily bitten in the past by the unnecessarily usage of cooperative threads, see https://github.com/zephyrproject-rtos/zephyr/pull/46172

carlocaione commented on dts/bindings/ipm/linaro,ivshmem-ipm.yaml:
```diff
@@ -0,0 +1,15 @@
+  ivshmem:
+    type: phandle
+    required: true
+    description: ivshmem driver node
```
`device node`

uLipe commented:
@carlocaione PTAL again when you have time.

gromero commented on samples/drivers/ipm/ipm_ivshmem/README.rst:
```diff
@@ -0,0 +1,129 @@
+QEMU source tree.
+
+ivshmem-client needs to be available as it is employed in this sample as an
+external application. The same conditions of ivshmem-server applies to the
```
s/applies/apply/ ?

gromero commented on samples/drivers/ipm/ipm_ivshmem/README.rst:
```diff
@@ -0,0 +1,129 @@
+
+Preparing IVSHMEM server
+************************
+#. The ivshmem-server utillity for QEMU can be found into the zephyr sdk
```
s/utillity/utility/ ?

gromero replied:
Zephyr SDK instead of zephyr sdk?

gromero commented on samples/drivers/ipm/ipm_ivshmem/README.rst:
```diff
@@ -0,0 +1,129 @@
+   directory, in:
+   ``/path/to/your/zephyr-sdk/zephyr-<version>/sysroots/x86_64-pokysdk-linux/usr/xilinx/bin/``
+
+#. You may also find ivshmem-client utillity, it can be useful to debug if everything works
```
s/utillity/utility/ ?

gromero replied:
Maybe s/to debug/to check/ since we debug things that are not working correctly?

gromero commented on samples/drivers/ipm/ipm_ivshmem/README.rst:
```diff
@@ -0,0 +1,129 @@
+#. Run ivshmem-server. For the ivshmem-server, both number of vectors and
+   shared memory size are decided at run-time (when the server is executed).
+   For Zephyr, the number of vectors and shared memory size of ivshmem are
+   decided at compile-time and run-time, respectively.For Arm64 we use
```
Missing space before "For Arm64..."?

gromero commented on samples/drivers/ipm/ipm_ivshmem/README.rst:
```diff
@@ -0,0 +1,129 @@
+   :goals: build
+   :compact:
+
+To run both qemu sides, repeat the west build command followed
```
s/qemu/QEMU/?

gromero reviewed (commented):
@uLipe Nice doc on IPM/ivshmem. Just found some nits you might want to fix. Cheers.

uLipe commented:
Just fixed some documentations nits, thanks @gromero
